modules/rostests/apitests/kernel32/UEFIFirmware.c

```diff
@@ -0,0 +1,221 @@
+    ok(dwLength == 0, "dwLength = %lu, expected 0\n", dwLength);
+    if (dwLength || dwError == ERROR_INVALID_FUNCTION)
+    {
+        skip("The rest of tests require UEFI environment, test abort.\n");
+        return;
+    }
+    /* Test ANSI version API too */
+    dwLengthTemp = GetFirmwareEnvironmentVariableA(EFI_DUMMY_VARIABLE_NAME,
+                                                   EFI_DUMMY_NAMESPACE_GUID_STRING,
+                                                   NULL,
+                                                   0);
+    dwErrorTemp = GetLastError();
+    ok(dwLengthTemp == dwLength && dwErrorTemp == dwError,
+       "dwLength = %lu, LastError = %lu, expected bResult = %lu, LastError = %lu\n",
+       dwLengthTemp,
+       dwErrorTemp,
+       dwLength,
+       dwError);
+
+    /* Generate a random as variable value to be used in this test */
+    RandomSeed = GetTickCount();
+    EfiVariableValue = RtlRandom(&RandomSeed);
+
+    /* Try to set firmware variable, should be failed with ERROR_PRIVILEGE_NOT_HELD,
+     * because no SE_SYSTEM_ENVIRONMENT_NAME privilege by default. */
+    bResult = SetFirmwareEnvironmentVariableW(_A2W(EFI_TEST_VARIABLE_NAME),
+                                              _A2W(EFI_TEST_GUID_STRING),
+                                              &EfiVariableValue,
+                                              sizeof(EfiVariableValue));
+    dwError = GetLastError();
+    ok(!bResult && dwError == ERROR_PRIVILEGE_NOT_HELD,
+       "bResult = %d, LastError = %lu, expected bResult = 0, LastError = ERROR_PRIVILEGE_NOT_HELD\n",
+       bResult,
+       dwError);
+    /* Test ANSI version API too */
+    bResultTemp = SetFirmwareEnvironmentVariableA(EFI_TEST_VARIABLE_NAME,
+                                                  EFI_TEST_GUID_STRING,
+                                                  &EfiVariableValue,
+                                                  sizeof(EfiVariableValue));
+    dwErrorTemp = GetLastError();
+    ok(bResultTemp == bResult && dwErrorTemp == dwError,
+       "bResult = %d, LastError = %lu, expected bResult = %d, LastError = %lu\n",
+       bResultTemp,
+       dwErrorTemp,
+       bResult,
+       dwError);
+
+    /* Enable SE_SYSTEM_ENVIRONMENT_NAME privilege that required by following tests */
+    bResult = OpenProcessToken(GetCurrentProcess(), TOKEN_ADJUST_PRIVILEGES, &hToken);
+    if (!bResult) {
+        skip("OpenProcessToken failed with error: 0x%08lX, test abort.\n", GetLastError());
+        return;
+    }
+    Privilege.PrivilegeCount = 1;
+    Privilege.Privileges[0].Luid = (LUID){ SE_SYSTEM_ENVIRONMENT_PRIVILEGE, 0 };
+    Privilege.Privileges[0].Attributes = SE_PRIVILEGE_ENABLED;
+    Status = NtAdjustPrivilegesToken(hToken, FALSE, &Privilege, sizeof(Privilege), NULL, &ReturnLength);
+    NtClose(hToken);
+    if (Status != STATUS_SUCCESS) {
+        skip("NtAdjustPrivilegesToken failed with status: 0x%08lX, test abort.\n", Status);
+        return;
```

Comment:
Shouldn't you restore the old privileges?

Reply:
In this code path, only one privilege was requested and fail, no privilege has been changed. So nothing to restore here.